cmake/multi_image.cmake

```diff
@@ -23,6 +23,259 @@ if(IMAGE_NAME)
+    string(REGEX MATCH "^${GET_ARGS_CHILD_IMAGE}_(.+)" unused_out_var ${var_name})
+
+    # When we try to pass a list on to the child image, like
+    # -DCONF_FILE=a.conf;b.conf, we will get into trouble because ; is
+    # a special character, so we escape it (mucho) to get the expected
+    # behaviour.
+    string(REPLACE \; \\\\\; val "${${var_name}}")
+
+    list(APPEND child_image_cmake_args
+      -D${CMAKE_MATCH_1}=${val}
+      )
+  endforeach()
+
+  unset(${GET_ARGS_CONFIG_OUT} PARENT_SCOPE)
+  set(${GET_ARGS_CONFIG_OUT} "${child_image_cmake_args}" PARENT_SCOPE)
+endfunction()
+
+function(add_s1_variant)
+  set(oneValueArgs BASE_IMAGE SOURCE_DIR)
+  cmake_parse_arguments(ASV "" "${oneValueArgs}" "" ${ARGN})
+
+  set(s1_override_conf
+    ${ZEPHYR_NRF_MODULE_DIR}/subsys/bootloader/image/is_s1_image.conf)
+
+  add_image_variant(
+    IMAGE_NAME s1_image
+    BASE_IMAGE ${ASV_BASE_IMAGE}
+    EXTRA_CONFIG ${s1_override_conf}
+    SOURCE_DIR ${ASV_SOURCE_DIR}
+    )
+endfunction()
+
+# Adds a variant of an image to the build.
+# The configuration passed to the variant image is a copy of that sent to the
+# base image, except what is passed to the EXTRA_CONFIG argument.
+#
+# Required arguments are:
+# IMAGE_NAME - The name of the variant image
+# BASE_IMAGE - The name of the base image which the variant image is based on.
+# SOURCE_DIR - The source dir of the base image.
+# EXTRA_CONFIG - The extra configuration fragment to pass to the variant image.
+function(add_image_variant)
+
+  set(oneValueArgs BASE_IMAGE IMAGE_NAME SOURCE_DIR EXTRA_CONFIG)
+  cmake_parse_arguments(AIV "" "${oneValueArgs}" "" ${ARGN})
+
+  if (AIV_BASE_IMAGE)
+    # A base image has been specified, this means that we need to copy the
+    # configuration of a child image (as opposed to copying the configuration
+    # from the top level parent image (that is, 'app').
+
+    set(base_image_configuration "${${AIV_BASE_IMAGE}_IMAGE_CMAKE_ARGS}")
+  else()
+    # No base image is specified, this means that the base image is the top
+    # level parent image (that is, 'app'). Copy relevant information from the
+    # 'app' image CMakeCache in order to build an identical variant image
+
+    get_cmake_property(variables_cached CACHE_VARIABLES)
+    foreach(var_name ${variables_cached})
+      # If '-DCONF_FILE' is specified, it is unset by boilerplate.cmake and
+      # replaced with 'CACHED_CONF_FILE' in the cache. Therefore we need this
+      # special handling for passing the value to the variant image.
+      if("${var_name}" MATCHES "CACHED_CONF_FILE")
+          list(APPEND application_vars ${var_name})
+      endif()
+
+      if("${var_name}" MATCHES "^CONFIG_.*")
+          list(APPEND application_vars ${var_name})
+      endif()
+
+      get_property(var_help CACHE ${var_name} PROPERTY HELPSTRING)
+      if("${var_help}" STREQUAL "No help, variable specified on the command line.")
+        list(APPEND application_vars ${var_name})
+      endif()
```

Comment:
Need a comment explaining this.

Reply:
Added comment in the top of the function